tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,291 @@
+            return false;
+        }
+
+        // Synchronize only if we are actually removing the span (stolen
+        // from std::Arc);
+        fence(Ordering::Acquire);
+        self.spans.remove(id_to_idx(&id))
+    }
+
+    fn drop_ref(&self, id: &Id) -> bool {
+        let idx = id_to_idx(id);
+        let span = self.spans.get(idx).expect("Span not found");
+
+        let refs = span.ref_count.fetch_sub(1, Ordering::Release);
+        assert!(
+            if std::thread::panicking() {
+                // don't cause a double panic, even if the ref-count is wrong...
+                true
+            } else {
+                refs != std::usize::MAX
+            },
+            "reference count overflow!"
+        );
+        refs == 1
+    }
+}
+
+thread_local! {
+    static CURRENT_SPANS: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+impl Subscriber for Registry {
+    fn register_callsite(&self, _: &'static Metadata<'static>) -> Interest {
+        Interest::always()
+    }
+
+    fn enabled(&self, _: &Metadata<'_>) -> bool {
+        true
+    }
+
+    #[inline]
+    fn new_span(&self, attrs: &span::Attributes<'_>) -> span::Id {
+        let mut values = HashMap::new();
+        let mut visitor = RegistryVisitor(&mut values);
+        attrs.record(&mut visitor);
+        let parent = attrs.parent().map(|id| self.clone_span(id));
+        let s = Data {
+            metadata: attrs.metadata(),
+            parent,
+            children: vec![],
+            ref_count: AtomicUsize::new(1),
+            values: Mutex::new(values),
+        };
+        let id = self.insert(s).expect("Unable to allocate another span");
+        idx_to_id(id)
+    }
+
+    #[inline]
+    fn record(&self, _: &span::Id, _: &span::Record<'_>) {
+        // TODO(david)—implement this.
+    }
+
+    fn record_follows_from(&self, _: &span::Id, _: &span::Id) {
+        // TODO(david)—implement this.
+    }
+
+    fn event(&self, event: &Event<'_>) {
+        let id = match event.parent() {
+            Some(id) => Some(id.clone()),
+            None => {
+                if event.is_contextual() {
+                    CURRENT_SPANS.with(|spans| spans.borrow().current().map(|id| id.clone()))
+                } else {
+                    None
+                }
+            }
+        };
+        if let Some(id) = id {
+            let mut values = HashMap::new();
+            let mut visitor = RegistryVisitor(&mut values);
+            event.record(&mut visitor);
+            let _ = self.get(&id).expect("Missing parent span for event");
+        }
+    }
+
+    fn enter(&self, id: &span::Id) {
+        CURRENT_SPANS.with(|spans| {
+            spans.borrow_mut().push(id.clone());
+            self.clone_span(id);
+        })
+    }
+
+    fn exit(&self, id: &span::Id) {
+        let _ = CURRENT_SPANS.with(|spans| spans.borrow_mut().pop(id));
+        dispatcher::get_default(|dispatch| dispatch.try_close(id.clone()));
```

Comment:
nit/style: you wouldn't need to clone the ID here if you just used the return value from the pop.

Reply:
also, we should probably only drop the ID if it was popped, in case someone called `exit` with a wrong ID. Like:
```rust
if let Some(id) = CURRENT_SPANS.with(|spans| spans.borrow_mut().pop(id)) {
    dispatcher::get_default(|dispatch| dispatch.try_close(id));
}
```
(or something; feel free to make it less ugly)

Reply:
Something like this should work:

```rust
if let Some(id) = CURRENT_SPANS.with(|spans| spans.borrow_mut().pop(id)) {
    dispatcher::get_default(|dispatch| dispatch.try_close(id.clone()));
}
```

Reply:
The `id` needs to be cloned in the scope of `get_default` because its an `FnMut`.

Reply:
ooh, that's fine then. you could also `take` it but this is probably better